### ISSUE TITLE
Inline string arguments appear to be unescaped twice

### ISSUE THREAD

**Issue by mattbaker** [absinthe-graphql/absinthe#961]
### Environment

* Elixir version (elixir -v): `Elixir 1.10.1 (compiled with Erlang/OTP 22)`
* Absinthe version (mix deps | grep absinthe): tested against `1.5.0-rc.4` and `master`
* Client Framework and version (Relay, Apollo, etc): N/A

### Expected behavior

Given a query with an inline string argument like so:

```
{
  field(argument: "\\\\FOO")
}
```

We would expect the resolver to receive the string `\\FOO`.

### Actual behavior

The resolver will receive `\FOO`.

Similarly, providing `\\\\\\\\FOO` as the argument yields `\\FOO` in the resolver, but we expected `\\\\FOO`.

It's worth mentioning that expected matches actual in Absinthe v1.4.16, we just never noticed the change after the v1.5 upgrade until now.

### Relevant Schema/Middleware Code

* `Absinthe.Lexer`
* `:absinthe_parser.yrl`
* `Absinthe.Phase.Parse`

### Details

@jmschles and I have been looking into some odd string escaping behavior, specifically with sequences of many `\`'s. As mentioned above, given a query with an inline string argument like so:

```
{
  field(argument: "\\\\FOO")
}
```

We would expect the resolver to receive the string `\\FOO` after the escape characters have been processed. Instead the resolver will receive `\FOO`.

A [lexer test](https://github.com/mattbaker/absinthe/commit/3fdd4609a4095da7abdb69d38a73dbe19cc254a6#diff-9ce8af4958da92722e4fba7e0e27e2ddR30-R50) shows the lexer is performing as expected, producing two literal slashes after tokenization:

```elixir
@query ~S"""
{
  { foo(bar: "\\\\FOO") }
}
"""
test "multiple escaped slashes"
  assert Absinthe.Lexer.tokenize(@query) ==
            {:ok,
            [
              {:"{", {1, 1}},
              {:"{", {2, 3}},
              {:name, {2, 5}, 'foo'},
              {:"(", {2, 8}},
              {:name, {2, 9}, 'bar'},
              {:":", {2, 12}},
              {:string_value, {2, 14}, ~S("\\FOO") |> String.to_char_list()},
              {:")", {2, 23}},
              {:"}", {2, 25}},
              {:"}", {3, 1}}
            ]}
end
```

The odd behavior becomes visible in an integration test. This will fail, for example ([full test file](https://github.com/mattbaker/absinthe/commit/3fdd4609a4095da7abdb69d38a73dbe19cc254a6#diff-04845d4af9c4d35dca26a712beb547a5R1-R72)):

```elixir
test "four slashes" do
  assert Absinthe.run(
            ~S"""
            {
              echo(value: "\\\\FOO")
            }
            """,
            Schema
          ) == {:ok, %{data: %{"echo" => ~S"\\FOO"}}} # <-- result of `run` will be ~S"\FOO"
end
```

As far as I can tell the unexpected behavior arises in `Absinthe.Phase.Parse.parse/1`. Specifically the call to `:absinthe_parser.parse/1` that occurs following tokenization by the lexer. 

```elixir
defp parse(input) do
    try do
      case tokenize(input.body) do # <-- Lexer runs
        {:ok, []} ->
          {:ok, %Language.Document{}}

        {:ok, tokens} ->
          case :absinthe_parser.parse(tokens) do # <-- Parser runs
            {:ok, _doc} = result ->
              result

            {:error, raw_error} ->
              {:error, format_raw_parse_error(raw_error)}
          end

        other ->
          other
      end
    rescue
      error ->
        {:error, format_raw_parse_error(error)}
    end
  end
```

Prior to #503 the parser handled escape sequences. The Nimble Parsec lexer (beautiful piece of work!) was introduced in #503 and [also handles escape sequences](https://github.com/absinthe-graphql/absinthe/blob/master/lib/absinthe/lexer.ex#L161-L183). While `:absinthe_parser` was updated as part of #503, the code to process escape sequences wasn't removed from the parser.

I _think_ the parser is redundantly processing escape sequences. We didn't notice because it's an odd edge case, essentially one round of escape processing by the lexer has to produce a string token that looks like another opportunity for escape processing by the parser. This is why other escape sequences like `\\n` yield [results I didn't expect](https://github.com/mattbaker/absinthe/commit/3fdd4609a4095da7abdb69d38a73dbe19cc254a6#diff-04845d4af9c4d35dca26a712beb547a5R62-R71).

As a test I commented out the `process_string/2` function heads that match on escape sequences in `absinthe_parser.yrl`. With that change I was able to get our new tests to pass, and existing tests continued to pass. _I don't know if this is right_, I'm pretty unfamiliar with this chunk of the code base, so I wanted to create this issue instead of propose a fix since my analysis could be wrong.

If it is true that we're processing escape sequences redundantly, I'd advocate for it being the responsibility of the lexer.

**Comment by bruce** [comment]
@mattbaker Thanks for reporting this. I have to do some investigation, but if it's true double processing is occurring, I agree with your assessment that the lexer should handle it (in the very least because that code is easier to reason about).

**Comment by mattbaker** [comment]
Awesome, and let me know if I can help somehow!

**Comment by binaryseed** [comment]
Fixed in #962
